### ISSUE TITLE
Feedback on logic error in code (possible)

### ISSUE THREAD

**Issue by CunminZhao** [med-air/3DSAM-adapter#15]
Feedback on logic error in code (possible): Hello, in this section of your code, the criterion for judging points_torch_negative is seg<10, which appears to be incorrect. In the dataset you provided, after processing, seg only has two values 0 and 1. Therefore, I think this section should be modified to seg<1.
“”“
l = len(torch.where(seg == 1)[0])
points_torch = None
if l > 0:
    sample = np.random.choice(np.arange(l), 40, replace=True)
    x = torch.where(seg == 1)[1][sample].unsqueeze(1)
    y = torch.where(seg == 1)[3][sample].unsqueeze(1)
    z = torch.where(seg == 1)[2][sample].unsqueeze(1)
    points = torch.cat([x, y, z], dim=1).unsqueeze(1).float()
    points_torch = points.to(device)
    points_torch = points_torch.transpose(0,1)
l = len(torch.where(seg < 10)[0])
sample = np.random.choice(np.arange(l), 10000, replace=True)
x = torch.where(seg < 10)[1][sample].unsqueeze(1)
y = torch.where(seg < 10)[3][sample].unsqueeze(1)
z = torch.where(seg < 10)[2][sample].unsqueeze(1)
points = torch.cat([x, y, z], dim=1).unsqueeze(1).float()
points_torch_negative = points.to(device)
points_torch_negative = points_torch_negative.transpose(0, 1)
”“”

**Comment by peterant330** [comment]
> Feedback on logic error in code (possible): Hello, in this section of your code, the criterion for judging points_torch_negative is seg<10, which appears to be incorrect. In the dataset you provided, after processing, seg only has two values 0 and 1. Therefore, I think this section should be modified to seg<1. “”“ l = len(torch.where(seg == 1)[0]) points_torch = None if l > 0: sample = np.random.choice(np.arange(l), 40, replace=True) x = torch.where(seg == 1)[1][sample].unsqueeze(1) y = torch.where(seg == 1)[3][sample].unsqueeze(1) z = torch.where(seg == 1)[2][sample].unsqueeze(1) points = torch.cat([x, y, z], dim=1).unsqueeze(1).float() points_torch = points.to(device) points_torch = points_torch.transpose(0,1) l = len(torch.where(seg < 10)[0]) sample = np.random.choice(np.arange(l), 10000, replace=True) x = torch.where(seg < 10)[1][sample].unsqueeze(1) y = torch.where(seg < 10)[3][sample].unsqueeze(1) z = torch.where(seg < 10)[2][sample].unsqueeze(1) points = torch.cat([x, y, z], dim=1).unsqueeze(1).float() points_torch_negative = points.to(device) points_torch_negative = points_torch_negative.transpose(0, 1) ”“”

This means if a sample does not have a foreground, then I will randomly sample points from any part of the image.  Actually <1 and <10 will give the same result. (We use <10 because we also do some experiments on muti-label segmentation)

**Comment by CunminZhao** [comment]
> > Feedback on logic error in code (possible): Hello, in this section of your code, the criterion for judging points_torch_negative is seg<10, which appears to be incorrect. In the dataset you provided, after processing, seg only has two values 0 and 1. Therefore, I think this section should be modified to seg<1. “”“ l = len(torch.where(seg == 1)[0]) points_torch = None if l > 0: sample = np.random.choice(np.arange(l), 40, replace=True) x = torch.where(seg == 1)[1][sample].unsqueeze(1) y = torch.where(seg == 1)[3][sample].unsqueeze(1) z = torch.where(seg == 1)[2][sample].unsqueeze(1) points = torch.cat([x, y, z], dim=1).unsqueeze(1).float() points_torch = points.to(device) points_torch = points_torch.transpose(0,1) l = len(torch.where(seg < 10)[0]) sample = np.random.choice(np.arange(l), 10000, replace=True) x = torch.where(seg < 10)[1][sample].unsqueeze(1) y = torch.where(seg < 10)[3][sample].unsqueeze(1) z = torch.where(seg < 10)[2][sample].unsqueeze(1) points = torch.cat([x, y, z], dim=1).unsqueeze(1).float() points_torch_negative = points.to(device) points_torch_negative = points_torch_negative.transpose(0, 1) ”“”
> 
> This means if a sample does not have a foreground, then I will randomly sample points from any part of the image. Actually <1 and <10 will give the same result. (We use <10 because we also do some experiments on muti-label segmentation)

thanks for your reply, I'm sorry, I was mistaken. I incorrectly took your "points_torch_negative" to be the points used to indicate background in SAM.

**Comment by csp2002** [comment]
Hi, I am still a little bit confused. In your code, it seems like if a sample has a foreground, you still concate the point and the negative point together. Why do we still need the negative point when we already have a point prompt?

**Comment by peterant330** [comment]
> Hi, I am still a little bit confused. In your code, it seems like if a sample has a foreground, you still concate the point and the negative point together. Why do we still need the negative point when we already have a point prompt?

This is part of noisy training that aims to improve the robustness of the model towards inaccurate prompts. The negative points are used only during the training phase.
